issue-40-built-in-benchmarking-tool.md

```diff
@@ -0,0 +1,31 @@
+## Add Built-in Benchmarking Tool
+
+## Summary
+
+There is a need to support a built-in benchmarking tool in Submariner.
+
+## Proposal
+
+Add a new ability to have a built-in benchmark tool.
+Implement a new Subctl command that will use iperf3 internally to benchmark the throughput between the two nodes.
+
+The feature Should support scheduling server and client in the same cluster.
+
+The user can specify source and destination nodes via node names. (Or via lables?)
+
+If nodes are not specified, the following will be the default tests: 
+* non-gateway to non-gateway  
+* gateway to gateway
+
+
+#### Source code location
+
+Source code should go into Submariner. 
+There is going to be a dependency on Shipyard (iperf image)
+
+
+### Subctl command
+
+```
+subctl test throughput --server-config, --server-node, --client-config and --client-node
```

Comment:
Please include kubeconfigs for both the clusters.

Reply:
`--server-config` and `--client-config` specify the kubeconfigs for both clusters.

Reply:
@skitt are there any parameters (e.g TCP Window) we need to expose at the "subctl test throughput" level or are we assuming that everything goes into the config files?

Reply:
Even in the datapath-connectivity tests, we kind of spawn a server/client (i.e., listener pod, connector pod), so it makes sense to align them for consistency.

The current command is as follows. So, if we decide to do the change, it will not require any changes to client scripts as well and would just require changes to subctl alone.

subctl verify \<kubeConfig1\> \<kubeConfig2\> [flags]

Reply:
> `--server-config` and `--client-config` specify the kubeconfigs for both clusters.

If --server-config/--client-config is for kubeconfig's, how do we provide the server/client options (like protocol/bitrate/time/windowSize etc)? as config-file or as command-line-args?

Reply:
> @skitt are there any parameters (e.g TCP Window) we need to expose at the "subctl test throughput" level or are we assuming that everything goes into the config files?

I'm also curious about how a user will be able to adjust the iperf3 parameters and what the defaults will be.
Will the benchmark loop through different number of parallel client streams by default?

Reply:
As per my comment on line 22, I don't think it makes sense to expose any of the tool internals to the user. If we want to change the internal benchmarking tool later, exposing these options as API either binds us to this implementation, or forces us to jump through loops or break the API.

Reply:
May be it's easier to not provide lots of options here as a start, and define a standardized testing, 

Like gw to gw and non-gw to non-gw to start with. Then we can start thinking about how to add fine tuning options, or the option to test in-cluster to add a baseline (may be we can include that part from the automated testing?)

Reply:
I have a slightly different opinion on the scope of the tool.

Personally, I feel that "subctl test throughput" is more like a wrapper/facilitator to schedule Pods on required nodes and run the performance tests with whatever configuration user provides and display the output as-is to the user.
The job of the tool is NOT to interpret the results or recommend any changes, it would be the job of the user running the tool.

So "subctl test throughput" would take the following inputs:
1. The backend tool that would be used to run the tests with default tool as netperf (or whatever we decide)
2. label of node where the Server/Client has to be scheduled
3. Server Options: This can be read from a configuration file and will be passed "as-is" to the respective tool while running the server side of the tool.
   Example: If "-s -1" is included in the configuration file, then iperf would be executed as "iperf -s -1"
4. Similarly client options: as above.

Each tool (like iperf or netperf) can display the output in its own manner. After the execution of the tests, subctl will simply output the same to the user. What are your thoughts on this approach?

Reply:
@maayanf24 Okay, based on the discussion we had yesterday, please identify the options that can be used for validating throughput and latency to begin with and we can expand the options (in future) based on feedback.